Text

LiamRiddell opened the pull request.

SebastianStehle commented:
Are you going to update the tests?

LiamRiddell commented:
> Are you going to update the tests?

Yes, I'm going to update them all.

LiamRiddell commented:
@SebastianStehle, Can you check the comments on this, please? 😄

SebastianStehle commented:
Which comments do you mean?

It is a little bit more work to get it done because of the new code generator, but the ConsoleApp1 can help you with that. I have also removed the resources file but it was annoying to work with and added tests files as embedded files. You can just copy and paste a file and then replace the content.

SebastianStehle commented on Mjml.Net/Helpers/ConditionalTags.cs:
```diff
@@ -0,0 +1,12 @@
+﻿namespace Mjml.Net.Helpers;
```
Just a minor thing: We should use a consistent syntax for namespaces.

LiamRiddell replied:
Agreed. However, see my comment below regarding changes to this file.

LiamRiddell commented:
> Which comments do you mean?
> 
> It is a little bit more work to get it done because of the new code generator, but the ConsoleApp1 can help you with that. I have also removed the resources file but it was annoying to work with and added tests files as embedded files. You can just copy and paste a file and then replace the content.

They should be in the files tab. I wrote then as inline comments on the code! Unless you can't see them?

SebastianStehle commented:
I cannot see any comment.

LiamRiddell commented:
> I cannot see any comment.

Okay give me a second I'll take pictures!

LiamRiddell commented:
![image](https://user-images.githubusercontent.com/3812154/158475633-5c5ef7d2-e808-40c1-b642-c21f286804ec.png)

LiamRiddell commented:
![image](https://user-images.githubusercontent.com/3812154/158475706-731c9d4d-ce02-423d-91e5-c14aa01ad514.png)

LiamRiddell commented:
> I cannot see any comment.

Silly me, I just realised I had not clicked "Submit Review"!! :D

SebastianStehle commented on Mjml.Net/Components/Body/TextComponent.cs:
```diff
@@ -0,0 +1,102 @@
+            }
+            else
+            {
+                renderer.Content(ConditionalTags.StartConditional);
```
It is great. Can you also change the other places?

LiamRiddell replied:
Apologies, ambiguity, which part is great? I personally believe we move this into renderer with the following syntax:

```csharp
MjmlRendererContext {
// ...
    void StartConditionalTag(EConditionalTagTargets conditionalTargets = EConditionalTagTargets.InternetExplorerAndMso, bool excludeTagTargets = false)
//...
}
```

SebastianStehle replied:
If you want to make it easier I would just introduce a few extension methods to the renderer to make the renderer as slim as possible. I also do not see why we need parameters, I mean there is no really logic there. It is always the same within a renderer. 

But if call it StartXXX and EndXXX we should also rename ElementStart to StartElement

SebastianStehle replied:
With other parts I mean the other components, like my components for example.

LiamRiddell replied:
@SebastianStehle yeah, I didn't like it in the end. I've already put my proposed solution in!

SebastianStehle commented on Mjml.Net/Components/Body/TextComponent.cs:
```diff
@@ -0,0 +1,102 @@
+                .Style("color", node.GetAttribute("color"))
+                .Style("height", node.GetAttribute("height"));
+
+            renderer.RenderChildren(new ChildOptions
```
What happens with ReadOuterXml()?

SebastianStehle replied:
I had to make a few changes to the context to fix some bugs. I would just focus on the update to the code generator and then have a look to this again. Perhaps the world looks different now.

SebastianStehle replied:
In general I think the RawXML should render everything, we would have the same problem in other places as well, I guess. For example mj-raw could have the same issue.

LiamRiddell replied:
> In general I think the RawXML should render everything, we would have the same problem in other places as well, I guess. For example mj-raw could have the same issue.

Exactly, early testing shows that `ReadOuterXml()` works. That's a bit of a confusing name :D I'll write some more unit tests. Agreed, this needs to be re-usable with minimal impact to render speed. I think maybe creating new function called `GetContentRaw()` would be more suitable. Otherwise, we'll be doing the conditional check for every `GetContentCall()`.

LiamRiddell commented:
@SebastianStehle, I've submitted the code but I'm getting build errors with the new generated properties. This affects other components too e.g. Hero, Spacer. 

```
The type 'Mjml.Net.Components.Body.TextProps' cannot be used as type parameter 'T' in the generic type or method 'BodyComponentBase<T>'. There is no boxing conversion from 'Mjml.Net.Components.Body.TextProps' to 'Mjml.Net.IProps'.	Mjml.Net

```

SebastianStehle commented:
It is not a build error, just a Visual Studio issue. If you have solved all errors except this one it should build fine.

LiamRiddell commented:
> It is not a build error, just a Visual Studio issue. If you have solved all errors except this one it should build fine.

Gotcha, so I did find one error which I thought was causing the build failure. I've fixed it and I'm still getting the below errors:

![image](https://user-images.githubusercontent.com/3812154/158486207-1affffbd-6bb6-4338-a021-6ca62df4f036.png)

SebastianStehle commented:
You have to pass props to RenderTextContent

LiamRiddell commented:
I see, sorry! This threw me off! It's getting late :D

```csharp
public abstract class BodyComponentBase<T> : BodyComponentBase where T : struct, IProps
    {
        public override IProps Props { get; } = default(T);
    }
   ```

SebastianStehle commented:
The props implement IProps over the code generator. But Visual Studio does not recognize it. Therefore the error list in VS makes often no sense. I will see how I can fix it, perhaps with a default implementation of the IProps methods or so.

LiamRiddell commented:
> The props implement IProps over the code generator. But Visual Studio does not recognize it. Therefore the error list in VS makes often no sense. I will see how I can fix it, perhaps with a default implementation of the IProps methods or so.
> 
> The output window is more reliable.

Yeah, now that I've experienced it. It won't be a problem again!

LiamRiddell commented:
Okay, I'm done for this component. If you could review that would be great! :)

SebastianStehle commented on Mjml.Net/Components/Body/TextComponent.cs:
```diff
@@ -0,0 +1,124 @@
+            }
+        }
+
+        private void RenderTextContent(IHtmlRenderer renderer, INode node, TextProps props)
```
I think it is faster to pass it over as "ref"

LiamRiddell replied:
You're right. Thank you!

SebastianStehle commented on Mjml.Net/Extensions/HtmlRendererExtensions.cs:
```diff
@@ -0,0 +1,61 @@
+        /// <summary>
+        /// Renders the start of conditional tag to NOT target Internet Explorer and Mso.
+        /// </summary>
+        /// <param name="renderer">The Html renderer to output too</param> 
```
There are a few warnings.

SebastianStehle commented on Mjml.Net/Components/Body/TextComponent.cs:
```diff
@@ -0,0 +1,124 @@
+
+            var rawContent = node.GetContentRaw();
+
+            renderer.Content(string.IsNullOrEmpty(rawContent) ? node.GetContent() : rawContent);
```
I don't like this, but if it works...I struggled so much with XmlReader as well :(

SebastianStehle replied:
Nothing to do here, I just wanted to mention that it. It should be solved in the ReadChildren I think.

LiamRiddell replied:
Yeah, essentially, if the Content Raw is empty we can assume the contents is `Text` else we know it's potentially `Html/Xml/Text`.

SebastianStehle reviewed (changes requested):
There are only very minor things, but it looks great now :)
